waspc/data/Generator/templates/sdk/wasp/server/operations/wrappers.ts

```diff
@@ -0,0 +1,233 @@
+  // the function as not accepting a payload (which isn't consistent with how
+  // it's defined).
+  return operation as UnauthenticatedOperationFor<OperationDefinition>
+}
+
+{=# isAuthEnabled =}
+// PRIVATE API (used in SDK)
+// Explanation:
+// - Custom `_Awaited` and `_ReturnType` - Read the comments above their
+// definitions.
+// - `Parameters<OperationDefinition> extends []` - Same reason as described here:
+// https://github.com/wasp-lang/wasp/pull/1992/files#r1583040080
+/**
+ * Constructs the authenticated operation's server-side API type from its
+ * definition.
+ *
+ * @template OperationDefinition The type of the authenticated operation's
+ * definition.
+ */
+export type AuthenticatedOperationFor<
+  OperationDefinition extends GenericOperationDefinition
+> = Parameters<OperationDefinition> extends []
+  ? AuthenticatedOperation<void, _Awaited<_ReturnType<OperationDefinition>>>
+  : AuthenticatedOperation<
+      Parameters<OperationDefinition>[0],
+      _Awaited<_ReturnType<OperationDefinition>>
+    >
+
+/**
+ * The type of the context users must pass when calling an authenticated
+ * operation's server-side API.
+ */
+export type AuthenticatedOperationContext = { user: AuthUser }
+
+// PRIVATE API (used in SDK)
+/**
+ * Creates the server-side API for an authenticated operation.
+ *
+ * @template OperationDefinition The type of the authenticated operation's definition.
+ * @param userOperation The authenticated operation's definition.
+ * @param entities The authenticated operation's entity map .
+ * @returns The server-side API for the provided authenticated operation.
+ */
+export function createAuthenticatedOperation<
+  OperationDefinition extends GenericOperationDefinition
+>(
+  userOperation: OperationDefinition,
+  entities: EntityMapFor<OperationDefinition>
+): AuthenticatedOperationFor<OperationDefinition> {
+  async function operation(...args: AuthenticatedOperationArgsFor<OperationDefinition>) {
+    // To understand this function and its limitations, read
+    // https://github.com/wasp-lang/wasp/issues/2050
+    if (args.length < 1) {
+      // No arguments sent -> no user and no payload specified -> there's no way this was called correctly.
+      throw new Error('Invalid number of arguments')
+    } else if (includesPayload(args)) {
+      // Two arguments sent -> the first argument is the payload, the second is the context.
+      const [payload, context] = args
+      return userOperation(payload, {
+        ...context,
+        entities,
+      })
+    } else {
+      // One argument sent -> the first and only argument is the user.
+      const [context] = args
+      return userOperation(undefined as Parameters<OperationDefinition>[0], {
+        ...context,
+        entities,
+      })
+    }
+  }
+
+  return operation as AuthenticatedOperationFor<OperationDefinition>
+}
+
+/**
+ * Returns a boolean carrying compiler type information about whether the
```

Comment:
This isn't an implementation detail, this is behaviour. Not mentioning implementation details in the top header means you are not explaining HOW it does what it does. But stuff like complexity (time, memory), assumptions it relies on, those are all part of the WHAT, now HOW, therefore part of API of this function. So not a really specific case, but an assumption, and those should be documented so the caller of the function can know what they can expect.

Reply:
So... is that a "yes" or a "no" to my suggestion? :smile:

Reply:
That means YES to the fact that you should document assumption in the header of the function and NO to the argument you made that this is some kind of special case, at least not the way you put it there.

I took a look, looks fine, although it is a bit out of context.

>  * To understand why "two arguments" means "includes payload", read
>  * https://github.com/wasp-lang/wasp/issues/2050
 
-> sounds like I already have context on what this "two arguments" is, but I am hearing about it for the first time. You are assuming I read the implementation I believe, but Imight have not and maybe I am just looking at the docs of this function, not how it is implemented.

So better would be something like this:

```
NOTE: This function assumes that payload is included if and only if there are 2 arguments. To understand why is this so, check ... .
```

Reply:
> You are assuming I read the implementation I believe

Not the implementation - the type signature, which is a part of the function's docs (check the screenshot above).

There's a "best practice" in typescript to not repeat type information in documentation (well, any static language, I guess :smile:).

In this case, the type signature says:
>  _This function takes a single argument. This argument is either an array with the payload and the context, or an array with just the context. It returns a type guard that says which one it is._ 

Once you know this, the implementation is straightforward and unambiguous (not that you need to know it). Therefore, this assumption
> NOTE: This function assumes that payload is included if and only if there are 2 arguments.

is made somewhere else (this function doesn't assume that at all, the caller does).

This function scope is simply, "If you give me something that's either A or B, I'll tell you which is it." The reasons why we're assuming A and B are the only options is within the caller's scope, and is documented there.

Reply:
I didn't maybe pay enough deep attention to that sentence, but neither will typical reader.

That sentence says it will receive an array with payload and context, or an array with just the context, and it will correctly determine which one it is. That doesn't so obviously translte into "it if it array with any two elems in it it will be one, or it is one elem it will be the other". And I remember you were saying before that we can't be smart enough here, that we are only checking if it has 1 or 2 elements.

Maybe I got something wrong here, but maybe it also doesn't matter -> what happened is that you had me read this, and I was confused. You possibly explained it to me here, but ideally your code would do that for you. So can you improve this to make it not confuse the next person, even though it is clear to you?